doc/design/blob-uuid-1.md

```diff
@@ -0,0 +1,51 @@
+A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
+
+** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.
+
+**key generation function**: the current suggestion (from satra) for this function is UUID4.
+
+**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.
+
+## Upload (from a CLI)
+
+### Check process (for de-duplication):
+1. CLI `GET /blobs/` <-- expected digest with digest type (e.g., ETag). API:
+    1. Looks for an existing AssetBlob with the given digest(s).
+    2. If it exists, return the key (UUID).
+    3. It it doesn't exist, return 404.
+2. If 404, start upload process.
+3. if key, call `POST .../assets/` with the key to register the new asset.
+
+### Blob upload process:
+1. CLI `/uploads/initialize` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:
```

Comment:
my only comment here is that we will want `contentSize` here and in the remaining places, since that's in the asset metadata. also the current digest doesn't look like above, but hopefully we will make that adjustment soon.

Reply:
`file_size` is here, which is a synonym for `contentSize`.
I typo'd the `algorithm`, it should be fixed now.

Reply:
we have contentSize in the schema, blobs are data blobs so might even have no file (at least with a name) so I would stick with `contentSize` without adding synonyms.